### ISSUE TITLE
Dynamic dependencies

### ISSUE THREAD

**Issue by Tishka17** [reagento/dishka#9]
Create something depending on, at least, request data

**Comment by Tishka17** [comment]
Need to add a way to understand what dependencies can be accessed during dynamic selection

**Comment by Skorpyon** [comment]
```python
class Dep:
    def __init__(self):
        self.dep_id = id(self)

    def __str__(self) -> str:
        return f'{self.__class__.__name__}({self.dep_id})'


class DependencyGlobal(Dep):
    ...


class Dependency1(Dep):
    ...


class Dependency2(Dep):
    ...


class Dependency3(Dep):
    ...


class MyProvider(Provider):
    # scope = Scope.REQUEST

    d1 = provide(DependencyGlobal, scope=Scope.APP)

    def __init__(self, scope=None):
        super().__init__(scope)

        for dep in [Dependency1, Dependency2, Dependency3]:
            setattr(self, str(dep.__name__), provide(dep, scope=Scope.REQUEST))

        self._init_dependency_sources()
```

Right now the only way is do some like that.

**Comment by Tishka17** [comment]
@Skorpyon I guess this will help: https://github.com/reagento/dishka/pull/61
It is not released but I hope to do it pretty soon

**Comment by Tishka17** [comment]
Version 0.4.0 released

**Comment by Skorpyon** [comment]
Absolutely amazing.

**Comment by Tishka17** [comment]
Another idea:
```py
@provide
def xxx(self, a: Deferred[XxxA], b: Deferred[XxxB], r: Request) -> XxxProto:
    if r.something:
       return a.ensure()
    else:
       return b.ensure()
```

**Comment by Tishka17** [comment]
One more alternative - selectors to determine which provider should be used

```
@provide(when=Has(Request))
def get_fastapi_user(self) -> User:
    return ...

@provide(when=Has(TelegramObject))
def get_tg_user(self, event: TelegramObject) -> User:
    return user_from_tg(event.from_user)
```

**Comment by Tishka17** [comment]
We agreed to have a predicate system which are checked after collecting all depedency keys.

The signature is `Callable[[BuildContext], bool]`, where BuildContext is a structure with field:
* root `context` passed to make_contaienr
* available keys
* current processing key with filled component

**Comment by Tishka17** [comment]
## Proposal: unified activation

### Basics

Activation is a function like factory or decorator, registered separately in provider

* Activation has a **Marker** which is used in other objects to check it using `when=` parameter 
* Activation function returns bool: is marker active or not.
* Activation function can have any dependencies as a normal provider or decorator
* Activation function result is available as `Activation[marker]` dependency
* Default activation always returns true

```python
class MyProvider(Provider):
    @activation("marker")
    def foo(self, settings: Settings) -> bool:
        return settings.x_type
    
    @provide(when="marker")
    def make_a(self) -> Proto:
        return A()
```

### Mutiple markers

Single activation function can be applied for multiple markers specified in decorator.
To access current value of a marker special type `Marker` is used. It behaves like a separate type in runtime and `Any` for type checkers.

```python
class MyProvider(Provider):
    @activation("A", "B")
    def foo(self, marker: Marker, settings: Settings) -> bool:
        return settings.x_type == marker
    
    @provide(when="A")
    def make_a(self) -> Proto:
        return A()

    @provide(when="B")
    def make_a(self) -> Proto:
        return B()
```

### Generic activation

Activation function can be bound not to concrete values but for specific type.


```python
@dataclass
class MyMarker:
    value: str

    
class MyProvider(Provider):
    @activation(type=MyMarker)
    def foo(self, marker: MyMarker, settings: Settings) -> bool:
        return settings.x_type == marker
    
    @provide(when=MyMarker("a"))
    def make_a(self) -> Proto:
        return A()

    @provide(when=MyMarker("b"))
    def make_a(self) -> Proto:
        return B()
```


### Static vs dynamic

Though it is not set explicitly, each activation function can be **static** or **dynamic**

* Static activation function is called only once when graph is built. All inactive providers are stripped from graph
* Dynamic activation function is called each time we need to decide which provider should be used.

If a dependency has multiple dynamically activated providers they are moved to a temporary component, while new factory is generated. 
This factory will call all activation functions consequently and once it gets True - requests dependency from corresponding factory. 

Rules to detect is it dynamic or not:
* async functions are always dynamic
* sync function with no arguments is static
* sync function which requests objects available in Root context is static
* sync function which requests Object which is not registered in graph is static
* sync function dependent on result of other static activation functions is static
* all other functions are dynamic


### Has(type) activator

Sometimes we need to check if there is a dependency.
It should be implemented as a prebuilt activation function registered for `Has(type)` objects.

* If provided `type` is registered as a factory/alias with a static activation, value from that activation is used
* If provided `type` is registered as `from_context` value with Scope equal to start container scope, it is considered as static activation and result is whether value is provided
* If provided `type` is registered as factory with dynamic activation or is context value with non-starting scope, then it works as a dynamic activation checking if corresponding type can be created

When checking if `type` as available it should not be instantiated.

**Comment by Tishka17** [comment]
Updated version with more unification

### Terms

1. **Activation** is a way to determine if dependency source (factory/alias/decorator) is used or not while being listed in Provider. 

2. **Activation function**/**Activator** is a special function that checks a condition used for dependency source activation.

3. **Activation Marker** is a special object used to bind dependency source with activator. 

### Basics


* Activator is a function like factory or decorator, registered separately in provider. 
* When registering, Activator is bound to a Marker or Marker class.
* Activator returns True if Marker is active, or False
* Each dependency source except context variables can be registered with `when=` parameter containing Marker instance. If Marker is active, dependency source is used. 
* Activator can have any dependencies as a normal provider or decorator
* `Activation[Marker]` can be requested to use activator result as a function parameter. 
* Dependency source without Activation Marker is considered always active 

```python
class MyProvider(Provider):
    @activation(Marker("a"))
    def foo(self, settings: Settings) -> bool:
        return settings.x_type
    
    @provide(when=Marker("a"))
    def make_a(self) -> Proto:
        return A()
```

### Mutiple markers

Single activation function can be applied for multiple markers specified in decorator.
To access current value of a marker special type `Marker` is used. It behaves like a separate type in runtime and `Any` for type checkers.

```python
class MyProvider(Provider):
    @activation(Marker("a"), Marker("b"))
    def foo(self, marker: Marker, settings: Settings) -> bool:
        return settings.x_type == marker.value
    
    @provide(when=Marker("a"))
    def make_a(self) -> Proto:
        return A()

    @provide(when=Marker("b"))
    def make_a(self) -> Proto:
        return B()
```

### Generic activation

Activation function can be bound not to concrete values but for specific Marker type.


```python
class MyMarker(Marker):
    pass

    
class MyProvider(Provider):
    @activation(MyMarker)
    def foo(self, marker: MyMarker, settings: Settings) -> bool:
        return settings.x_type == marker
    
    @provide(when=MyMarker("a"))
    def make_a(self) -> Proto:
        return A()

    @provide(when=MyMarker("b"))
    def make_a(self) -> Proto:
        return B()
```

### Marker combination

You can combining Marker objects to have more complex condition

```python

    @provide(when=~Marker("b") | Marker("c"))
    def make_a(self) -> Proto:
        return B()
```

### Static vs dynamic

Though it is not set explicitly, each activation function can be **static** or **dynamic**

* Static activator is called only once when graph is built. All inactive providers are stripped from graph
* Dynamic activator is called each time we need to decide which provider should be used.

If a dependency has multiple dynamically activated providers they are moved to a temporary component, while new factory is generated. 
This factory will call all activators consequently and once it gets True - requests dependency from corresponding factory. 

Rules to detect is it dynamic or not:
* **async** functions are always **dynamic**
* sync function with **no arguments** is **static**
* sync function which requests objects available in **Root context** is **static**
* sync function which requests Object which is **not registered** in graph is **static**
* sync function dependent on result of other **static** activators is **static**
* all other functions are **dynamic**


### Has(type) activator

`Has` is a subclass of `Marker` intended for specific use with prebuild activator.

Sometimes we need to check if there is a dependency.
It should be implemented as a prebuilt activation function registered for `Has(type)` objects.

* If provided `type` is registered as a factory/alias with a static activation, value from that activation is used
* If provided `type` is registered as `from_context` value with Scope equal to start container scope, it is considered as static activation and result is whether value is provided
* If provided `type` is registered as factory with dynamic activation or is context value with non-starting scope, then it works as a dynamic activation checking if corresponding type can be created

When checking if `type` as available it should not be instantiated. 

### Activator scope

Activator result is cached within scope. Scope is determined based on scopes of dependencies and requester.

1. If there is multiscoped dependency (Container or AsyncContainer for now), than scope of requester is used
2. In other cases, the most nested scope of all objects is used

Hereby, only Activators with scope=Runtime (or custom analogue) can be used as static.

### Activator with both dynamic/static behavior

Some activators can be treated as both dynamic and static based on Marker value.
In this case we need delegate type detection to activator itself. This needs to be clarified.

**Comment by Tishka17** [comment]
### Components and activators

Each marker can be used in different components separately. By default, `when=` condition is checked only against activators set in the same component as current factory (regardles we used `FromComponent` as a provide hint). 

At the same time we can want to access activators in another component. Let's say we have Component "A" with activator
```python
@activate(MarkerClassA)
def foo(marker: MarkerClassA) -> bool: ...
```
In another component we have factory activated using
```python
@provide(when=MarkerClassA("x"))
```
The proposal is to allow aliases refer to markers in another components limiting that we cannot alias one type to different type. 
```python
alias(MarkerClassA("x"), component="A")  # ok
alias(MarkerClassA, component="A")  # ok
alias(MarkerClassA, OtherMarkerClass)  # not ok
alias(Marker("a"), Has(int)) # ok again
```

**Comment by Tishka17** [comment]
https://dishka.readthedocs.io/en/latest/advanced/when.html
